New docs structure (#188)
iterative/mlem.ai: content/docs/get-started/index.md

# Get Started

We assume MLEM is already [installed](/doc/install) in your active Python

> environment, as well as `pandas` and `sklearn`. If not, you can follow the
> instructions below:

Reviewer: 💅🏼 
```suggestion
environment, as well as `pandas` and `sklearn` (if not, you can follow the
instructions below).
```

<details>

### ⚙️ Expand for setup instructions

Let's create a separate folder and an isolated virtual environment to cleanly

> install all the requirements (including MLEM) there:

Reviewer: 💅🏼 
```suggestion
install all the requirements (including MLEM):
```

```cli
$ mkdir mlem-get-started
$ cd mlem-get-started
$ python3 -m venv .venv

> $ source .venv/bin/activate

Reviewer: Linux/Mac-specific.

Do we expect Windows users? Maybe use tabs if so ([example](https://dvc.org/doc/start/data-management/data-versioning#expand-to-set-up-remote-storage)).

Same for other commands (e.g. `tree` and pages in the GS).

Reviewer 2: as per analytics, only 3% of MLEM command calls are from windows. So let's skip it for now

$ pip install pandas scikit-learn mlem[fastapi,heroku]
```

</details>

## Saving your ML model

> To enable all kinds of productionization scenarios MLEM supports, we first need
> to save a ML model with MLEM.

Reviewer: ```suggestion
To enable all kinds of productionization scenarios supported by MLEM, we first need
to save a machine learning model with MLEM.
```

Let's take a look at the following Python script:

```py
# train.py
from sklearn.datasets import load_iris
from sklearn.ensemble import RandomForestClassifier

from mlem.api import save

data, y = load_iris(return_X_y=True, as_frame=True)
rf = RandomForestClassifier(
    n_jobs=2,
    random_state=42,
)
rf.fit(data, y)
save(
    rf,
    "models/rf",
    sample_data=data,
)
```

> Here we load well-known iris dataset with sklearn and train a simple classifier.
> But instead of pickling the model we saved it with MLEM (check out the
> [full list of ML frameworks](/doc/object-reference/model) that MLEM supports.)

Reviewer: ```suggestion
Here, we load a well-known
[Iris flower dataset](https://archive.ics.uci.edu/ml/datasets/iris) with scikit-learn
and train a simple classifier. But instead of pickling the model, we save it with
MLEM (check out the full list of supported
[ML frameworks](/doc/object-reference/model)).
```

> Now let's run this script and see how MLEM saved the model.

Reviewer: ```suggestion
Let's run this and see the results in the project directory:
```

```cli
$ python train.py
...

$ tree models/
models
├── rf
└── rf.mlem
```

> The model was saved along with some metadata about it: `rf` is the model binary
> and `rf.mlem` is a metafile containing information about the model. Let's take a
> look at it:

Reviewer: ```suggestion
The model binary was saved to `models/rf`, along with some metadata about it
in `models/rf.mlem`.
```

<details>

> ### `$ cat models/rf.mlem`

Reviewer: ```suggestion
### Click to see the contents of the `rf.mlem` metafile.
```

```yaml
artifacts:
  data:
    hash: 5a38e5d68b9b9e69e9e894bcc9b8a601
    size: 163651
    uri: rf
model_type:
  methods:
    predict:
      args:
        - name: data
          type_:
            columns:
              - sepal length (cm)
              - sepal width (cm)
              - petal length (cm)
              - petal width (cm)
            dtypes:
              - float64
              - float64

>               - float64

Reviewer: Do we need all the lines? Even if this is hidden, it's super long. Is it informative at all? Can we truncate some parts to highlight what matters most?

Reviewer 2: Good point, made it shorter a bit.

      args:
        - name: X
          type_:
            columns:
              - sepal length (cm)
              - sepal width (cm)
              - petal length (cm)
              - petal width (cm)
            dtypes:
              - float64
              - float64
              - float64
              - float64
            index_cols: []
            type: dataframe
      name: predict
      returns:
        dtype: int64
        shape:
          - null
        type: ndarray
    sklearn_predict_proba:
      args:
        - name: X
          type_:
            columns:
              - sepal length (cm)
              - sepal width (cm)
              - petal length (cm)
              - petal width (cm)
            dtypes:
              - float64
              - float64
              - float64
              - float64
            index_cols: []
            type: dataframe
      name: predict_proba
      returns:
        dtype: float64
        shape:
          - null
          - 3
        type: ndarray
  type: sklearn
object_type: model
requirements:
  - module: sklearn
    version: 1.1.2
  - module: numpy
    version: 1.22.4
  - module: pandas
    version: 1.5.0
```

</details>

> It's a bit long, but we can see all that we need to use the model later:

Reviewer: ```suggestion
The `.mlem` file is a bit long, but it contains all the metadata we need to use
the model later:
```

1. Model methods: `predict` and `predict_proba`

> 2. Input data schema that describes the DataFrame with the iris dataset
> 3. Requirements: `sklearn`, `pandas` with particular versions we need to run
>    this model.

Reviewer: ```suggestion
2. Input data schema: describes the data frame (Iris dataset)
3. Requirements: `sklearn` and `pandas`, with specific versions
```

> Note that we didn't specify requirements: MLEM investigates the object you're
> saving (even if it's a complex one) and finds out all requirements needed!

Reviewer: ```suggestion
Note that we didn't specify any of this information. MLEM investigates the object
(even if it's complex) and finds out all of this!
```

## Productionization

> Saving models is fun, but the real value of a model is how you can use it. To
> make it easier to get models to production, MLEM has 4 related functionalities:
> applying, serving, building and deploying.

Reviewer: ```suggestion
Saving ML models is fun, but what we really want to do is to use them! To make
it easier to get models to production, MLEM has 4 strategies: applying, serving,
building, and deploying.
```

> - **[Applying models](/doc/get-started/applying)** explains how to load model in
>   Python or get prediction in command line.

Reviewer: ```suggestion
- **[Applying models](/doc/get-started/applying)** means loading them with
  Python, or getting predictions directly from command line (system terminal).
```

> - **[Serving models](/doc/get-started/serving)** shows how to create a service
>   from your model.

Reviewer: ```suggestion
- **[Serving models](/doc/get-started/serving)** wraps them as a web interface
  (e.g. a REST API).
```

> - **[Exporting models (building)](/doc/get-started/building)** describes how
>   models can be built into Python packages, Docker images, etc.

Reviewer: ```suggestion
- **[Exporting models](/doc/get-started/building)** is achieved by building
  them as Python packages, Docker images, etc. that you can easily distribute.
```

> - **[Deploying models](/doc/get-started/deploying)** shows how you can deploy
>   your model with MLEM.

Reviewer: ```suggestion
- **[Deploying models](/doc/get-started/deploying)** to cloud platforms
  (SageMaker, Kubernetes, etc.) helps you make it available for others to
  utilize easily and reliably.
```

More examples on how to use MLEM in different scenarios can be found in

> [Use Cases](/doc/use-cases) section.

Reviewer: I think we can remove this text.

# Get Started

We assume MLEM is already [installed](/doc/install) in your active Python
environment, as well as `pandas` and `sklearn`. If not, you can follow the
instructions below:

<details>

### ⚙️ Expand for setup instructions

Let's create a separate folder and an isolated virtual environment to cleanly
install all the requirements (including MLEM) there:

```cli
$ mkdir mlem-get-started
$ cd mlem-get-started
$ python3 -m venv .venv
$ source .venv/bin/activate

> $ pip install mlem pandas sklearn fastapi docker

Author: ```suggestion
$ pip install mlem pandas sklearn fastapi uvicorn docker
```

```

</details>

## Saving your ML model

To enable all kinds of productionization scenarios MLEM supports, we first need
to save a ML model with MLEM.

Let's take a look at the following Python script:

```py
# train.py
from sklearn.datasets import load_iris
from sklearn.ensemble import RandomForestClassifier

from mlem.api import save

data, y = load_iris(return_X_y=True, as_frame=True)
rf = RandomForestClassifier(
    n_jobs=2,
    random_state=42,
)
rf.fit(data, y)
save(
    rf,
    "models/rf",
    sample_data=data,
)
```

Here we load well-known iris dataset with sklearn and train a simple classifier.

> But instead of pickling the model we saved it with MLEM. (Check out the

Author: ```suggestion
But instead of pickling the model we saved it with MLEM (check out the
```

[full list of ML frameworks](/doc/object-reference/model) that MLEM supports.)

> Now let's run this script and see how we save the model.

Author: ```suggestion
Now let's run this script and see how the model was saved.
```

```cli
$ python train.py
...

$ tree models/
models
├── rf
└── rf.mlem
```

> The model was saved along with some metadata about it: `rf` containing the model

Author: ```suggestion
The model was saved along with some metadata about it: `rf` is the model
```

> binary and `rf.mlem` metafile containing information about it. Let's take a look

Author: ```suggestion
binary and `rf.mlem` is a metafile containing information about the model. Let's take a look
```
